rpc/providers/proxy.ts

```diff
@@ -3,44 +3,38 @@ import { deadline, deferred } from "../../deps/std/async.ts";
 import { ErrorCtor } from "../../util/mod.ts";
 import * as B from "../Base.ts";
 import { ClientHooks, ParseRawIngressMessageError } from "../common.ts";
+import { WebSocketProxy } from "./web-socket-proxy.ts";
 
 export type ProxyClientHooks<Config_ extends Config<string>> = ClientHooks<Config_, Event>;
 
-export function proxyClient<Config_ extends Config<string>>(
+export async function proxyClient<Config_ extends Config<string>>(
   config: Config_,
   hooks?: ProxyClientHooks<Config_>,
 ): Promise<ProxyClient<Config_> | FailedToOpenConnectionError> {
-  const ws = new WebSocket(config.discoveryValue);
-  const client = new ProxyClient(ws, hooks);
-  ws.addEventListener("error", client.onError);
-  ws.addEventListener("message", client.onMessage);
-  const pending = deferred<ProxyClient<Config_> | FailedToOpenConnectionError>();
-  if (ws.readyState === WebSocket.CONNECTING) {
-    const onOpenError = (e: any) => {
-      console.log({ log: e });
-      clearListeners();
-      pending.resolve(new FailedToOpenConnectionError());
-    };
-    const onOpen = () => {
-      clearListeners();
-      pending.resolve(client);
-    };
-    const clearListeners = () => {
-      ws.removeEventListener("error", onOpenError);
-      ws.removeEventListener("open", onOpen);
-    };
-    ws.addEventListener("error", onOpenError);
-    ws.addEventListener("open", onOpen);
-  } else {
-    pending.resolve(client);
+  const wsIsOpen = deferred();
+  const ws = new WebSocketProxy({
+    webSocketFactory: () => new WebSocket(config.discoveryValue),
+  });
+
+  ws.onopen = () => wsIsOpen.resolve();
+  ws.onerror = () => wsIsOpen.reject();
+
+  try {
+    await wsIsOpen;
+  } catch (_e) {
+    return new FailedToOpenConnectionError();
+  } finally {
+    ws.onopen = undefined;
+    ws.onerror = undefined;
   }
-  return pending;
+
+  return new ProxyClient(ws, hooks);
 }
 
 export class ProxyClient<Config_ extends Config>
   extends B.Client<Config_, MessageEvent, Event, FailedToDisconnectError>
 {
-  constructor(ws: WebSocket, hooks?: ProxyClientHooks<Config_>) {
+  constructor(ws: WebSocketProxy, hooks?: ProxyClientHooks<Config_>) {
```

Comment:
Right now, hooks are not used.
And there is a mapping between a message request and a successful/failed response.

However, when the WebSocket connection is closed, the inflight requests are not completed with an error (or I'm not able to spot this in the current Client implementation).

Perhaps, this could be handed by the `Client`, Thoughts?

Reply:
+1 to removing hooks. The hook-in approach isn't nearly as ergonomic as separating inflight message / dispatch state into a higher-level container that holds onto the `WsContainer` (or whatever we end up calling it). So yes: I think your approach makes a lot of sense: `Client` should hold onto the given state; perhaps through a `ClientState` / something reusable across providers. Then, retrying is simply about detaching from the provider (in this case `WsContainer`) and ensuring the new callbacks are hooked up to the existing provider.